pinot-segment-local/src/main/java/org/apache/pinot/segment/local/dedup/PartitionDedupMetadataManager.java

```diff
@@ -114,12 +146,13 @@ public PrimaryKey next() {
   }
 
   public boolean checkRecordPresentOrUpdate(PrimaryKey pk, IndexSegment indexSegment) {
-    boolean present =
-        _primaryKeyToSegmentMap.putIfAbsent(HashUtils.hashPrimaryKey(pk, _hashFunction), indexSegment) != null;
-    if (!present) {
+    byte[] keyBytes = serializePrimaryKey(HashUtils.hashPrimaryKey(pk, _dedupConfig.getHashFunction()));
+    if (Objects.isNull(_keyValueStore.get(keyBytes))) {
```

Comment:
This is a non-atomic operation (contrast with `putIfAbsent`).

Reply:
Will fix this by bringing back the putIfAbsent.